sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferStore.scala

```diff
@@ -304,11 +305,48 @@ abstract class RapidsBufferStore(
       }
     }
 
+    override def getMemoryBuffer: MemoryBuffer = materializeMemoryBuffer
+
+    override def getDeviceMemoryBuffer: DeviceMemoryBuffer = {
+      deviceRapidsBuffer match {
+        case Some(rapidsBuffer) => rapidsBuffer.getDeviceMemoryBuffer
+        case None =>
+          catalog.acquireBuffer(id, DEVICE) match {
+            case Some(acquiredBuffer) =>
+              deviceRapidsBuffer = Some(acquiredBuffer)
+              deviceRapidsBuffer.get.getDeviceMemoryBuffer
+            case _ =>
+              val newBuffer = {
+                logDebug(s"Unspilling $this $id to $DEVICE")
+                RapidsBufferCatalog.getDeviceStorage.copyBuffer(this, Cuda.DEFAULT_STREAM)
+              }
+              try {
+                catalog.registerNewBuffer(newBuffer)
+              } catch {
+                case e: IllegalStateException =>
+                  // A device buffer is already registered.
+                  logDebug(s"Error registering new buffer $newBuffer", e)
+                  newBuffer.free()
+                  deviceRapidsBuffer = catalog.acquireBuffer(id, DEVICE)
+                  return deviceRapidsBuffer.get.getDeviceMemoryBuffer
```

Comment:
This needs to loop back around and try to acquire from the beginning to cover the case where it still returns `None`.  Yes, should be incredibly rare in practice, but theoretically possible if this thread gets indefinitely postponed.

Reply:
Done.